patches/server/0844-Collision-API.patch

```diff
@@ -0,0 +1,48 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
+Date: Wed, 6 Oct 2021 20:10:44 -0400
+Subject: [PATCH] Collision API
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java b/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java
+index fe7d73470773179f8254928f8e3cdc19f64e9dd7..78eedd85819c2583a0b42a94c3e9e4bacbdd8a51 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java
+@@ -894,4 +894,13 @@ public abstract class CraftRegionAccessor implements RegionAccessor {
+ 
+         throw new IllegalArgumentException("Cannot spawn an entity for " + clazz.getName());
+     }
++
++    // Paper Start - Collision API
++    @Override
++    public boolean collidesWith(@org.jetbrains.annotations.NotNull org.bukkit.util.BoundingBox boundingBox) {
++        net.minecraft.world.phys.AABB aabb = new AABB(boundingBox.getMinX(), boundingBox.getMinY(), boundingBox.getMinZ(), boundingBox.getMaxX(), boundingBox.getMaxY(), boundingBox.getMaxZ(), false);
++
++        return !this.getHandle().noCollision(aabb);
++    }
++    // Paper End - Collision API
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+index 0bd800e1aeda87689a6c56ee6fadda381c74a4ff..819d2166498d15461c85ff2188be66c95ca5ca3a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+@@ -1287,4 +1287,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+         return !entity.valid && entity.level.addEntity(entity, reason);
+     }
+     // Paper end
++    // Paper Start - Collision API
++    @Override
++    public boolean collidesAt(@org.jetbrains.annotations.NotNull Location location) {
++        net.minecraft.world.phys.AABB aabb = this.getHandle().getBoundingBoxAt(location.getX(), location.getY(), location.getZ());
++
++        return !this.getHandle().level.noCollision(this.getHandle(), aabb);
++    }
++
++    @Override
++    public boolean collidesAt(@org.jetbrains.annotations.NotNull BoundingBox boundingBox) {
++        net.minecraft.world.phys.AABB aabb = new AABB(boundingBox.getMinX(), boundingBox.getMinY(), boundingBox.getMinZ(), boundingBox.getMaxX(), boundingBox.getMaxY(), boundingBox.getMaxZ(), false);
```

Comment:
In regards to repeated conversion between AABB and BoundingBox, the MCUtil utility class does not yet seem to have a neat conversion method for this.

Might be worth to add one [into this patch](https://github.com/PaperMC/Paper/blob/master/patches/server/0006-MC-Utils.patch) and use that to convert from/to an AABB.

Incredibly minor nitpick tho, so really up to you.

Reply:
Decided against this in the end, as really this constructor is only needed here once anyways... So IDK, and the constructor used in question is added in a later patch.